neural_compressor/compression/pruner/model_slim/pattern_analyzer.py

```diff
@@ -0,0 +1,381 @@
+        static_graph: The static graph of original model.
+        flatten_static_graph: A list of string with the model's static graph inference details.
+        target_layers: The layer types the searcher will extract.
+        searching_results: The list/dict which store matched patterns.
+        target_path: a string presenting the pattern to search (use '/' to link different layer names.)
+        target_op: a set containing target operators' names.
+        target_op_lut: a lookup table for target operators and their corresponding jit codes.
+        current_pattern: a searching path to store searching status.
+    """
+
+    def __init__(self, model, target_pattern='linear/gelu/linear'):
+        """Initialize."""
+        super(PathSearcher, self).__init__(model)
+        # some search related attribuites
+        self.target_pattern = target_pattern
+        # re-org target_pattern to obtain target ops
+        self.target_path = self.target_pattern.split('/')
+        self.target_ops = set(self.target_path)
+        self.target_op_lut = {}
+        self.current_pattern = []
+    
+    def search(self):
+        """Operations called for entire searching process."""
+        # step 1: establish search space within all interested ops, saved in self.target_op_lut
+        self.search_results.clear()
+        self.target_op_lut.clear()
+        self.current_pattern.clear()
+        for op in self.target_ops:
+            self.target_op_lut[op] = JitBasicSearcher.filter_static_code(self, self.flatten_static_graph, "aten::"+op)
+        def dfs():
+            """Operations called for one single search step."""
+            # if target pattern is found
+            if len(self.current_pattern) == len(self.target_path):
+                # find the target pattern
+                self.search_results.append(self.current_pattern[:])
+                return
+            # else continue searching
+            next_op_type = self.target_path[len(self.current_pattern)]
+            for op_code in self.target_op_lut[next_op_type]:
+                op_info = JitBasicSearcher.analyze_code(self, op_code)
+                if len(self.current_pattern) == 0: 
+                    self.current_pattern.append(op_info)
+                    dfs()
+                    self.current_pattern.pop()
+                else:
+                    if op_info['input_name'] == self.current_pattern[-1]['output_name']:
+                        self.current_pattern.append(op_info)
+                        dfs()
+                        self.current_pattern.pop()
+                    else:
+                        continue 
+        # step 2: dfs  
+        # execute dfs-based pattern matching
+        dfs()
+        print(f"Matched {len(self.search_results)} pattern {self.target_pattern} in model {type(self.model).__name__}")
+        JitBasicSearcher.get_layer_for_all(self)
+        return self.search_results
+
+class Linear2LinearSearcher(JitBasicSearcher):

```

Comment:
could we implement it with some pattern rather than a class

Reply:
Will do! All improvements including functions and coding structure will be carried out in next version.

Reply:
MHA slim will also be supported, since hf seems to have similar features.